File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-0.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        if (head == null) return null; 
        HashMap<Node, Node> map = new HashMap<>();
        Node newhead = head;
    
       
        while (newhead!= null){
            map.put(newhead, new Node(newhead.val));
            newhead= newhead.next;
        }
        newhead=head;
        while (newhead!=null){
            Node copy = map.get(newhead);
            copy.next= map.get(newhead.next);
            copy.random= map.get(newhead.random);
            newhead=newhead.next;
        }
        return map.get(head);
    }
}

File: Data Structures & Algorithms/populating-next-right-pointers-in-each-node/submission-0.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
public:
    Node* connect(Node* root) {
        queue<Node*> q;
        if(!root){
            return nullptr;
        }
        q.push(root);
        while(!q.empty()){
            int level=q.size();
            while(level>0){
                 Node* node = q.front();
                 q.pop();
                 if(level>1){
                    node->next= q.front();
                 }
                 if(node->left){
                    q.push(node->left);
                 }
                 if(node->right){
                    q.push(node->right);
                 }
                 level--;
            }
            
        }
        return root;
    }
};